tests/integrationv2/test_hello_retry_requests.py

```diff
@@ -0,0 +1,71 @@
+import copy
+import os
+import pytest
+import time
+
+from configuration import available_ports, TLS13_CIPHERS, ALL_TEST_CURVES, ALL_TEST_CERTS
+from common import ProviderOptions, Protocols, data_bytes, Curves
+from fixtures import managed_process
+from providers import Provider, S2N, OpenSSL
+from utils import invalid_test_parameters, get_parameter_name
+   
+# List of keyshares for hello retry requests client side test.
+HRR_CLIENT_KEYSHARES = [
+    ["-K", "none"],
+    ["-K", "secp256r1"],
+    ["-K", "secp256r1:secp384r1"],
+]
+
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", TLS13_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("provider", [OpenSSL])
+@pytest.mark.parametrize("curve", ALL_TEST_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", ALL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("keyshare", HRR_CLIENT_KEYSHARES, ids=get_parameter_name)
+def test_hrr_with_s2n_as_client(managed_process, cipher, provider, curve, protocol, certificate, keyshare):
+    port = next(available_ports)
+
+    random_bytes = data_bytes(64)
+    client_options = ProviderOptions(
+        mode=Provider.ClientMode,
+        host="localhost",
+        port=port,
+        cipher=cipher,
+        data_to_send=random_bytes,
+        insecure=True,
+        curve=curve,
+        extra_flags=keyshare,
+        protocol=protocol)
+
+    server_options = copy.copy(client_options)
+    server_options.data_to_send = None
+    server_options.mode = Provider.ServerMode
+    server_options.key = certificate.key
+    server_options.cert = certificate.cert
+    server_options.extra_flags = None
+    server_options.curve = Curves.X25519
+
+    # Passing the type of client and server as a parameter will
+    # allow us to use a fixture to enumerate all possibilities.
+    server = managed_process(provider, server_options, timeout=5)
+    client = managed_process(S2N, client_options, timeout=5)
+
+    # The client should connect and return without error
+    for results in client.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+        assert bytes("Curve: {}".format("x25519").encode('utf-8')) in results.stdout
+
+    marker_part1 = b"cf 21 ad 74 e5"
+    marker_part2 = b"9a 61 11 be 1d"
+
+    for results in server.get_results():
+        assert results.exception is None
```

Comment:
Previously we were validating that multiple ClientHello messages were passed, and multiple Finished messages were passed. I think we should still do that because we actually missed a bug where the retry wasn't working, but didn't verify multiple ClientHellos were involved.

Reply:
I was mistaken earlier, we can only validate that from the s_server results, in s_client "client hello" and "server hello" is not explicitly called out. Ex: 

```
Stdout: b'Using default temp DH parameters
ACCEPT
read from 0x5647dae89e80 [0x5647dae97713] (5 bytes => 5 (0x5))
0000 - 16 03 01 00 ec                                    .....
read from 0x5647dae89e80 [0x5647dae97718] (236 bytes => 236 (0xEC))
0000 - 01 00 00 e8 03 03 07 bd-f8 6a a9 02 d2 78 25 19   .........j...x%.
0010 - 7b 42 6e f3 33 52 b7 d5-5a 8c 96 61 0c d8 fa c2   {Bn.3R..Z..a....
0020 - b4 24 6d 14 63 1e 00 00-4e 00 04 00 05 00 0a 00   .$m.c...N.......
0030 - 16 00 2f 00 33 00 35 00-39 00 3c 00 3d 00 67 00   ../.3.5.9.<.=.g.
0040 - 6b 00 9c 00 9d 00 9e 00-9f 13 01 13 02 13 03 c0   k...............
0050 - 09 c0 0a c0 11 c0 12 c0-13 c0 14 c0 23 c0 24 c0   ............#.$.
0060 - 27 c0 28 c0 2b c0 2c c0-2f c0 30 cc a8 cc a9 cc   \'.(.+.,./.0.....
0070 - aa ff 04 ff 08 00 ff 01-00 00 71 00 2b 00 0b 0a   ..........q.+...
0080 - 03 04 03 03 03 02 03 01-03 00 00 33 00 02 00 00   ...........3....
0090 - 00 0d 00 26 00 24 08 09-08 0a 08 0b 08 04 08 05   ...&.$..........
00a0 - 08 06 04 01 05 01 06 01-03 01 04 03 04 03 05 03   ................
00b0 - 05 03 06 03 03 03 02 01-02 03 00 00 00 0e 00 0c   ................
00c0 - 00 00 09 6c 6f 63 61 6c-68 6f 73 74 00 0a 00 08   ...localhost....
00d0 - 00 06 00 1d 00 17 00 18-00 0b 00 02 01 00 fe 01   ................
00e0 - 00 0a 00 08 00 0d 00 13-00 01 00 0a               ............
TLS client extension "supported versions" (id=43), len=11
0000 - 0a 03 04 03 03 03 02 03-01 03 00                  ...........
TLS client extension "key share" (id=51), len=2
0000 - 00 00                                             ..
TLS client extension "signature algorithms" (id=13), len=38
0000 - 00 24 08 09 08 0a 08 0b-08 04 08 05 08 06 04 01   .$..............
0010 - 05 01 06 01 03 01 04 03-04 03 05 03 05 03 06 03   ................
0020 - 03 03 02 01 02 03                                 ......
TLS client extension "server name" (id=0), len=14
0000 - 00 0c 00 00 09 6c 6f 63-61 6c 68 6f 73 74         .....localhost
TLS client extension "supported_groups" (id=10), len=8
0000 - 00 06 00 1d 00 17 00 18-                          ........
TLS client extension "EC point formats" (id=11), len=2
0000 - 01 00                                             ..
write to 0x5647dae89e80 [0x5647daea0930] (67 bytes => 67 (0x43))
0000 - 16 03 03 00 38 02 00 00-34 03 03 cf 21 ad 74 e5   ....8...4...!.t.
0010 - 9a 61 11 be 1d 8c 02 1e-65 b8 91 c2 a2 11 16 7a   .a......e......z
0020 - bb 8c 5e 07 9e 09 e2 c8-a8 33 9c 00 13 01 00 00   ..^......3......
0030 - 0c 00 2b 00 02 03 04 00-33 00 02 00 1d 14 03 03   ..+.....3.......
0040 - 00 01 01                                          ...
read from 0x5647dae89e80 [0x5647dae97713] (5 bytes => 5 (0x5))
0000 - 14 03 03 00 01                                    .....
read from 0x5647dae89e80 [0x5647dae97718] (1 bytes => 1 (0x1))
0000 - 01                                                .
read from 0x5647dae89e80 [0x5647dae97713] (5 bytes => 5 (0x5))
0000 - 16 03 03 01 10                                    .....
read from 0x5647dae89e80 [0x5647dae97718] (272 bytes => 272 (0x110))
0000 - 01 00 01 0c 03 03 5d f0-ce 2b 1e 83 56 22 72 86   ......]..+..V"r.
0010 - dd b2 bc 83 f7 50 85 04-9f a5 d0 73 39 0a da b8   .....P.....s9...
0020 - 08 2f ce 95 94 9e 00 00-4e 00 04 00 05 00 0a 00   ./......N.......
0030 - 16 00 2f 00 33 00 35 00-39 00 3c 00 3d 00 67 00   ../.3.5.9.<.=.g.
0040 - 6b 00 9c 00 9d 00 9e 00-9f 13 01 13 02 13 03 c0   k...............
0050 - 09 c0 0a c0 11 c0 12 c0-13 c0 14 c0 23 c0 24 c0   ............#.$.
0060 - 27 c0 28 c0 2b c0 2c c0-2f c0 30 cc a8 cc a9 cc   \'.(.+.,./.0.....
0070 - aa ff 04 ff 08 00 ff 01-00 00 95 00 2b 00 0b 0a   ............+...
0080 - 03 04 03 03 03 02 03 01-03 00 00 33 00 26 00 24   ...........3.&.$
0090 - 00 1d 00 20 2a b9 5a f9-db 2d a9 58 b4 fa 3b 50   ... *.Z..-.X..;P
00a0 - 33 f4 f0 91 77 73 a6 44-9b a7 cd 44 b2 b8 29 f7   3...ws.D...D..).
00b0 - 3d d8 b6 74 00 0d 00 26-00 24 08 09 08 0a 08 0b   =..t...&.$......
00c0 - 08 04 08 05 08 06 04 01-05 01 06 01 03 01 04 03   ................
00d0 - 04 03 05 03 05 03 06 03-03 03 02 01 02 03 00 00   ................
00e0 - 00 0e 00 0c 00 00 09 6c-6f 63 61 6c 68 6f 73 74   .......localhost
00f0 - 00 0a 00 08 00 06 00 1d-00 17 00 18 00 0b 00 02   ................
0100 - 01 00 fe 01 00 0a 00 08-00 0d 00 13 00 01 00 0a   ................
TLS client extension "supported versions" (id=43), len=11
0000 - 0a 03 04 03 03 03 02 03-01 03 00                  ...........
TLS client extension "key share" (id=51), len=38
0000 - 00 24 00 1d 00 20 2a b9-5a f9 db 2d a9 58 b4 fa   .$... *.Z..-.X..
0010 - 3b 50 33 f4 f0 91 77 73-a6 44 9b a7 cd 44 b2 b8   ;P3...ws.D...D..
0020 - 29 f7 3d d8 b6 74                                 ).=..t
TLS client extension "signature algorithms" (id=13), len=38
0000 - 00 24 08 09 08 0a 08 0b-08 04 08 05 08 06 04 01   .$..............
0010 - 05 01 06 01 03 01 04 03-04 03 05 03 05 03 06 03   ................
0020 - 03 03 02 01 02 03                                 ......
TLS client extension "server name" (id=0), len=14
0000 - 00 0c 00 00 09 6c 6f 63-61 6c 68 6f 73 74         .....localhost
TLS client extension "supported_groups" (id=10), len=8
0000 - 00 06 00 1d 00 17 00 18-                          ........
TLS client extension "EC point formats" (id=11), len=2
0000 - 01 00     
```

Which is why now I am verifying the Shared Groups and the Supported Groups in the output.

Reply:
Ahh I see, nice.